src/Endatix.Infrastructure/Integrations/Slack/SlackClient.cs

```diff
@@ -19,28 +21,64 @@ public class SlackClient : INotificationHandler<SubmissionCompletedEvent>, ISlac
     private readonly ILogger<SlackClient> _logger;
     private readonly IRepository<Form> _formRepository;
     private readonly SlackSettings _slackSettings;
-    public SlackClient(ILogger<SlackClient> logger, IRepository<Form> formRepository, IOptions<SlackSettings> options) {
+    private readonly IServiceProvider _serviceProvider;
+    //TODO: The Slack message template should be configurable through the Hub's UI
+    // {0} is the submissionURL, {1} is the form name
+    private const string SLACK_MESSAGE_TEMPLATE = ":page_with_curl: <{0}|New submission> for {1}";
+    // Submission URL parameters: {0} is the Hub's base URL, {1} is FormId
+    private const string SLACK_SUBMISSIONS_URL_TEMPLATE = "{0}/forms/submissions/{1}";
+    public SlackClient(ILogger<SlackClient> logger,
+            IRepository<Form> formRepository,
+            IOptions<SlackSettings> options,
+            IServiceProvider serviceProvider)
+    {
         _formRepository = formRepository;
         _logger = logger;
         _slackSettings = options.Value;
+        _serviceProvider = serviceProvider;
     }
 
-    public async Task Handle(SubmissionCompletedEvent notification, CancellationToken cancellationToken) {
-        if(notification.Submission.IsComplete && _slackSettings.Active){
+    public async Task Handle(SubmissionCompletedEvent notification, CancellationToken cancellationToken)
+    {
+        Guard.Against.Null(_slackSettings.Active, nameof(_slackSettings.Active), "Slack settings must specify whether the integration is active.");
```

Comment:
It's good that you added checks for the settings.
Something that I wish to suggest is that these checks are done in the beginning of the constructor. This way we would ensure on successful creation of the SlackClient object, it has everything needed to do its job. In addition the checks would not be necessary in the different methods of the class.

Reply:
Done. Unfortunatelly now we get "possible null" warnings in the Handle function.

Reply:
Yes, there is a relatively new operator for such cases - putting ! after an object of nullable type explicitly says the value is not null, like e.g. `(bool)_slackSettings.Active!`. The warning disappears this way,

Reply:
Applied